### PR TITLE
Don't open twice the same activity

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
## Fixes
If you click twice clickly on a button which opens an activity, this activity will be opened twice. This is the case for example when clicking on a deck name to review it. When clicking on edit from the editor. When clicking on «add note» from the deck browser window. When clicking on «card browser» on the side menu...

This is a problem in particular when you edit notes, because when you ends editing the note, the activity close, and you'll see the previous activity, which is also card edition, with the data of the previous version of the note. If you simply cancel, nothing is saved, and the actual new version of the note is kept, but it's not clear at all to any normal user.

In other cases, it's not a huge problem, it's just slightly confusing to see again card to review when you clicked «back» in the reviewer. 

I should note that fixing the problem of registering two clicks instead of one should still be fixed; e.g. when you click twice on "undo" you may query "undo" twice, even if there is a single thing to undo. So that's only a partial solution to the real problem that I do in this PR.

## Approach
Intents use FLAG_ACTIVITY_SINGLE_TOP and FLAG_ACTIVITY_CLEAR_TOP to ensure that a single activity is used when we don't want two activities. I don't add those flags when either: 
* it calls an activity from outside of ankidroid (I believe that if we want to add a media, the media app will deal itself with the fact that it's opened twice)
* we actually want twice the same activity (the only example I have being in «Edit note» when we want to copy the current note. This open the same activity with a new note; and then we want to come back to the edited note.

More precisely, I created a class inheriting Intent, having those flags by default. This allows to avoid doing a copy paste where we add them in 16 files.

## How Has This Been Tested?

Installing ankidroid on my phone, and testing many of the features mentioned above. Trying to click twice clickly, and seeing I don't have the same bug as before.

## Learning (optional, can help others)
_Links to blog posts, patterns, libraries or addons used to solve this problem_
https://stackoverflow.com/a/44433420/414269

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 1]
```diff
@@ -1765,7 +1766,7 @@ private void restoreCollectionPreferences() {
             throw new RuntimeException();
         } catch (NullPointerException npe) {
             // NPE on collection only happens if the Collection is broken, follow AnkiActivity example
-            Intent deckPicker = new Intent(this, DeckPicker.class);
+            Intent deckPicker = new IntentTop(this, DeckPicker.class);
             deckPicker.putExtra("collectionLoadError", true); // don't currently do anything with this
             deckPicker.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java]: A great deal of these calls are followed by a FLAG_ACTIVITY_CLEAR_TOP being redundantly added now, those extra CLEAR_TOPs should be removed?

Also - as a maybe - quite a few also add FLAG_ACTIVITY_NEW_TASK (so I assume some decisions about tasks and task affinity are baked in). Perhaps after removing the redundant flags, a second Intent helper extending the first expressing (with a good name) the idea of single top + new task could be used.

**Comment by mikehardy** [comment]
This is a subtle problem and very weird (I imagine) if you trigger it. Good catch

**Comment by Arthur-Milchior** [comment]
As requested above, I removed the .addFlags which became useless. I also created another class having directly New_Task.
Concerning clear_task, I still left it as something added outside; does not seems to occur often enough to add it as yet another class

**Comment by mikehardy** [comment]
2.10alpha48 is published to google and will be available with this as soon as google is done processing 🚀

**Comment by mikehardy** [comment]
This was reverted just now, released with revert as 2.10alpha53

**Pull request merged**